spec/controllers/idv/by_mail/enter_code_controller_spec.rb

```diff
@@ -1,23 +1,84 @@
 require 'rails_helper'
 
 RSpec.describe Idv::ByMail::EnterCodeController do
-  let(:has_pending_profile) { true }
-  let(:success) { true }
-  let(:otp) { 'ABC123' }
-  let(:submitted_otp) { otp }
-  let(:user) { create(:user) }
-  let(:profile_created_at) { Time.zone.now }
-  let(:pending_profile) do
-    if user
-      create(
-        :profile,
-        :with_pii,
-        user: user,
-        proofing_components: proofing_components,
-        created_at: profile_created_at,
-      )
+  describe 'Pii::Cacher use' do
+    let(:params) { {} }
+    let(:session) { {} }
+    let(:pii_cacher) { Pii::Cacher.new(user, session) }
+
+    before do
+      allow(Pii::Cacher).to receive(:new).and_return(pii_cacher)
+      allow(pii_cacher).to receive(:fetch).and_call_original
+
+      stub_sign_in(user)
+    end
+
+    context 'when the user has no profiles' do
+      let(:user) { create(:user) }
+
+      describe '#index' do
+        before { get :index, params: params, session: session }
+
+        it 'uses no PII' do
+          expect(pii_cacher).not_to have_received(:fetch)
+        end
+      end
+
+      describe '#create' do
+        before { put :index, params: params, session: session }
+
+        it 'uses no PII' do
+          expect(pii_cacher).not_to have_received(:fetch)
+        end
+      end
+    end
+
+    context 'when the user has a verify by mail (GPO) pending profile' do
+      let(:user) { create(:user, :with_pending_gpo_profile) }
+      let(:pending_profile) { user.gpo_verification_pending_profile }
+
+      describe '#index' do
+        before { get :index, params: params, session: session }
+
+        it 'uses the PII from the pending profile' do
+          expect(pii_cacher).to have_received(:fetch).with(pending_profile.id)
+        end
+      end
+
+      describe '#create' do
+        before { put :index, params: params, session: session }
+
+        it 'uses the PII from the pending profile' do
+          expect(pii_cacher).to have_received(:fetch).with(pending_profile.id)
+        end
+      end
+    end
+
+    context 'when the user has an active profile, but no verify by mail (GPO) profile' do
+      let(:user) { create(:user, :fully_registered) }
+
+      describe '#index' do
+        before { get :index, params: params, session: session }
+
+        it 'uses no PII' do
+          expect(pii_cacher).not_to have_received(:fetch)
+        end
+      end
+
+      describe '#create' do
+        before { put :index, params: params, session: session }
+
+        it 'uses no PII' do
+          expect(pii_cacher).not_to have_received(:fetch)
+        end
+      end
     end
   end
+end
+
+RSpec.describe Idv::ByMail::EnterCodeController do
```

Comment:
It is kind of strange to me to see 2 `describe` blocks in here. Wdyt of moving the setup into the `#index` and `#create` blocks and combining these 2 top level describe blocks? I think there are some advantages to that anyway moving the setup closer to the actual tests.

Reply:
For example I don't see a reason that the `#index` tests need most of these values.

Reply:
Kept the describe blocks separate until I'd finished cleaning up the old one. Merge them.